File: Scripts/utils/Graphics/graphics_base.lua
local graphics = { }

function graphics.IsGraphicsLibraryLoaded()
    return Call("IsGraphicsLibraryLoaded", "boolean")
end

function graphics.GetAvailableFonts()
    Call("GetAvailableFonts", "nil")
end
function graphics.AddFont(fontFile, fontName, sizePixels)
    Call("AddFont", "nil string string float", fontFile, fontName, sizePixels)
end
function graphics.DrawLine(x1, y1, x2, y2, r, g, b, alpha, thickness)
    return Call("DrawLine", "number float float float float number number number number float", x1, y1, x2, y2, r, g, b, alpha, thickness)
end

function graphics.DrawRect(x, y, width, height, r, g, b, alpha, rounding, flags)
    return Call("DrawRect", "number float float float float number number number number float number", x, y, width, height, r, g, b, alpha, rounding, flags)
end

function graphics.DrawRectWithBorders(x, y, width, height, r, g, b, alpha, borderR, borderG, borderB, borderThickness, rounding, flags)
    return Call("DrawRectWithBorders", "number float float float float number number number number float number", x, y, width, height, r, g, b, alpha, borderR, borderG, borderB, borderThickness, rounding, flags)
end

function graphics.DrawGradientRect(x, y, width, height, left_bottom_red, left_bottom_green, left_bottom_blue, left_bottom_alpha, left_top_red, left_top_green, left_top_blue, left_top_alpha, right_top_red, right_top_green, right_top_blue, right_top_alpha, right_bottom_red, right_bottom_green, right_bottom_blue, right_bottom_alpha)
    return Call("DrawGradientRect", "number float float float float number number number number number number number number number number number number number number number number", x, y, width, height, left_bottom_red, left_bottom_green, left_bottom_blue, left_bottom_alpha, left_top_red, left_top_green, left_top_blue, left_top_alpha, right_top_red, right_top_green, right_top_blue, right_top_alpha, right_bottom_red, right_bottom_green, right_bottom_blue, right_bottom_alpha)
end

function graphics.DrawTriangle(x1, y1, x2, y2, x3, y3, r, g, b, alpha)
    return Call("DrawTriangle", "number float float float float float float number number number number", x1, y1, x2, y2, x3, y3, r, g, b, alpha)
end

function graphics.DrawEllipse(x, y, radiusX, radiusY, r, g, b, alpha, numSegments)
    return Call("DrawEllipse", "number float float float float number number number number number", x, y, radiusX, radiusY, r, g, b, alpha, numSegments)
end

function graphics.DisplayText(text, x, y, font, font_size, r, g, b, alpha)
    return Call("DisplayText", "number string float float string float number number number number", text, x, y, font, font_size, r, g, b, alpha)
end

function graphics.DrawWatermark(text, x, y, width, height, start_r, start_g, start_b, end_r, end_g, end_b, alpha, font, font_size)
    return Call("DrawWatermark", "number string float float float float number number number number number number number string float", text, x, y, width, height, start_r, start_g, start_b, end_r, end_g, end_b, alpha, font, font_size)
end

function graphics.DrawImage(image_path, x, y, width, height)
    return Call("DrawImage", "number string float float number number", image_path, x, y, width, height)
end

function graphics.ShowNotification(text, duration)
    Call("ShowNotification", "nil string number", text, duration)
end

function graphics.SetGlobalFontSize(font_size)
    Call("SetGlobalFontSize", "nil float", font_size)
end
function graphics.SetGlobalFontSizeStr(font_size)
    Call("SetGlobalFontSizeStr", "nil string", font_size)
end

function graphics.SetElementPosition(id, ...)
    Call("SetElementPosition", "nil number table", id, {...})
end

function graphics.SetElementSize(id, width, height)
    Call("SetElementSize", "nil number number number", id, width, height)
end

function graphics.SetElementColor(id, r, g, b, alpha)
    Call("SetElementColor", "nil number number number number number", id, r, g, b, alpha)
end

function graphics.SetElementText(id, text)
    Call("SetElementText", "nil number string", id, text)
end

function graphics.SetElementExtra(id, ...)
    Call("SetElementExtra", "nil number table", id, {...})
end

function graphics.SetElementFont(id, font, font_size)
    Call("SetElementFont", "nil number string float", id, font, font_size)
end

function graphics.DeleteElement(id)
    Call("DeleteElement", "nil number", id)
end

function graphics.ClearRenderList()
    Call("ClearRenderList", "nil")
end

function graphics.GetDisplaySize()
    local size = Call("GetDisplaySize", "table(2)")
    return size[1], size[2]
end

return graphics

--Call("DrawLine", "nil float float float float number number number number float", 0.0, 0.0, 300.0, 400.0, 255, 0, 0, 255, 1.0)
--Call("DrawRect", "nil float float float float number number number number float number", 1300.0, 90.0, 80.0, 385.0, 255, 255, 0, 255, 1.0, 0)
--Call("DrawTriangle", "nil float float float float float float number number number number", 1000.0, 1000.0, 1200.0, 1000.0, 1400.0, 1000.0, 0, 100, 199, 255)
--Call("DrawEllipse", "nil float float float float number number number number number", 500.0, 500.0, 70.0, 70.0, 0, 200, 140, 255, 30)
--Call("DisplayText", "nil string float float float number number number number", "hello", 100.0, 90.0, 50.0, 212, 107, 65, 255)
--Call("ShowNotification", "nil string number", "Executed by loader.lua", 5000)

File: Scripts/utils/Graphics/graphics_base.js
const Graphics = {
    isGraphicsLibraryLoaded: function () {
        return Call("IsGraphicsLibraryLoaded", "boolean");
    },

    getAvailableFonts: function () {
        Call("GetAvailableFonts", "nil");
    },

    addFont: function (fontFile, fontName, sizePixels) {
        Call("AddFont", "nil string string float", fontFile, fontName, sizePixels);
    },

    drawLine: function (x1, y1, x2, y2, r, g, b, alpha, thickness) {
        return Call("DrawLine", "integer float float float float number number number number float", x1, y1, x2, y2, r, g, b, alpha, thickness);
    },

    drawRect: function (x, y, width, height, r, g, b, alpha, rounding, flags) {
        return Call("DrawRect", "integer float float float float number number number number float number", x, y, width, height, r, g, b, alpha, rounding, flags);
    },

    drawRectWithBorders: function (x, y, width, height, r, g, b, alpha, borderR, borderG, borderB, borderThickness, rounding, flags) {
        return Call("DrawRectWithBorders", "integer float float float float number number number number float number", x, y, width, height, r, g, b, alpha, borderR, borderG, borderB, borderThickness, rounding, flags);
    },

    drawGradientRect: function(x, y, width, height) {
        // Получаем дополнительные аргументы начиная с позиции 4.
        var colors = Array.prototype.slice.call(arguments, 4);
    
        // Эмулируем repeat для строки "number " 16 раз (используем цикл)
        var protoNumbers = "";
        for (var i = 0; i < 16; i++) {
            protoNumbers += "number ";
        }
        
        var prototypeStr = "integer float float float float " + protoNumbers;
    
        // Формируем массив аргументов для Call:
        // Первый элемент – имя функции, второй – прототип, затем x, y, width, height и остальные аргументы
        var args = ["DrawGradientRect", prototypeStr, x, y, width, height].concat(colors);
    
        // Используем apply для передачи массива аргументов
        return Call.apply(null, args);
    },    

    drawTriangle: function (x1, y1, x2, y2, x3, y3, r, g, b, alpha) {
        return Call("DrawTriangle", "integer float float float float float float number number number number", x1, y1, x2, y2, x3, y3, r, g, b, alpha);
    },

    drawEllipse: function (x, y, radiusX, radiusY, r, g, b, alpha, numSegments) {
        return Call("DrawEllipse", "integer float float float float number number number number number", x, y, radiusX, radiusY, r, g, b, alpha, numSegments);
    },

    displayText: function (text, x, y, font, font_size, r, g, b, alpha) {
        return Call("DisplayText", "integer string float float string float number number number number", text, x, y, font, font_size, r, g, b, alpha);
    },

    drawWatermark: function (text, x, y, width, height, start_r, start_g, start_b, end_r, end_g, end_b, alpha, font, font_size) {
        return Call("DrawWatermark", "integer string float float float float number number number number number number number string float", text, x, y, width, height, start_r, start_g, start_b, end_r, end_g, end_b, alpha, font, font_size);
    },

    drawImage: function (imagePath, x, y, width, height) {
        return Call("DrawImage", "integer string float float number number", imagePath, x, y, width, height);
    },

    showNotification: function (text, duration) {
        Call("ShowNotification", "nil string number", text, duration);
    },

    setGlobalFontSize: function (fontSize) {
        Call("SetGlobalFontSize", "nil float", fontSize);
    },

    setGlobalFontSizeStr: function (fontSize) {
        Call("SetGlobalFontSizeStr", "nil string", fontSize)
    },

    setElementPosition: function (id) {
        var position = Array.prototype.slice.call(arguments, 1);
        Call("SetElementPosition", "nil number array", id, position);
    },

    setElementSize: function (id, width, height) {
        Call("SetElementSize", "nil number number number", id, width, height);
    },

    setElementColor: function (id, r, g, b, alpha) {
        Call("SetElementColor", "nil number number number number number", id, r, g, b, alpha);
    },

    setElementText: function (id, text) {
        Call("SetElementText", "nil number string", id, text);
    },

    setElementExtra: function (id) {
        var extra = Array.prototype.slice.call(arguments, 1);

        Call("SetElementExtra", "nil number array", id, extra);
    },

    setElementFont: function (id, font, fontSize) {
        Call("SetElementFont", "nil number string float", id, font, fontSize);
    },

    deleteElement: function (id) {
        Call("DeleteElement", "nil number", id);
    },

    clearRenderList: function () {
        Call("ClearRenderList", "nil");
    },

    getDisplaySize: function () {
        var size = Call("GetDisplaySize", "array(2)");

        return { width: size[0], height: size[1] };
    }
};

globalThis.Graphics = Graphics; // Export to global scope